lib/THCUNN/LogSoftMax.cu

```diff
@@ -1,6 +1,81 @@
 #include "THCUNN.h"
 #include "common.h"
 
+__global__ void cunn_SpatialLogSoftMax_updateOutput_kernel(float *output, float *input, int classSize, int height, int width)
+{
+  int batchIndex = blockIdx.x;
+  int index = threadIdx.x;
+
+  while (index < height*width) {
+    int y = index / width;
+    int x = index % width;
+    if (y >= height)
+      break;
+
+    // calculate input starting index in cuda layout (B x H x W x P)
+    int inputStartIndex = 0
+    + (height*width*classSize)*batchIndex
+    + (width*classSize)*y
+    + (classSize)*x;
+
+    // compute a
+    float a = 0;
+    for (int i = 0; i < classSize; i++) {
+      a += __expf(input[inputStartIndex + i]);
+    }
+
+    // invert a
+    a = 1.0f / a;
```

Comment:
This really doesn't need this kind of comment (same goes for `// compute a`) 😉 If you want to add some insights, it would be more helpful to say what a is (e.g. show it in a softmax formula). It's not a very good name though, it's not conventionally used anywhere in context of Softmax.